lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2861,6 +2861,62 @@ bool FoldChannelShuffle::run(Function *F, const CompilationContext &cctx) {
   return changed;
 }
 
+// Fold Tile -> Add into BatchedAdd wherever applicable.
+bool FoldTileAddIntoBatchedAdd::run(Function *F,
+                                    const CompilationContext &cctx) {
+  LOG_SCOPE(F->getLogContext(), getName());
+
+  bool changed = false;
+  auto &nodes = F->getNodes();
+  for (auto &node : nodes) {
+    auto *addNode = dyn_cast<AddNode>(&node);
+    if (!addNode) {
+      continue;
+    }
+
+    NodeValue batchNode, addedNode;
+    auto &LHS = addNode->getLHS();
+    auto &RHS = addNode->getRHS();
+    auto *tileNode = dyn_cast<TileNode>(LHS);
+
+    // Check if LHS is a tile.
+    if (tileNode) {
+      batchNode = RHS;
+      addedNode = tileNode->getInput();
+    }
+    // Check if RHS is a tile.
+    else if ((tileNode = dyn_cast<TileNode>(RHS))) {
+      batchNode = LHS;
+      addedNode = tileNode->getInput();
+      // If neither LHS or RHS is a tile, nothing to do.
+    } else {
+      continue;
+    }
+
+    // If the tiling is not along the 0th axis,
+    // Add cannot be replaced with BatchedAdd.
+    if (tileNode->getAxis() != 0) {
+      continue;
+    }
+    auto oldDims = addedNode.dims();
+    // If the 0th dimension is not 1, then reducing dimension
+    // via reshaping is more complicated. Hence, Add will not be
+    // replaed with BatchedAdd.
+    if (oldDims.size() < 2 || oldDims[0] != 1) {
```

Comment:
Why do you skip the optimization if `oldDims.size() < 2`? You should still be able to do it without reshaping if `oldDims.size() == 1` and `oldDims[0] == 1`, right?

Reply:
Done!